### PR TITLE
Add native support for range field types by using a range object

### LINKED ISSUE #1862: Add native support for range field types by using a range object
Currently there is no native support for range field types [1], except the one explained in the blog post [2]. However, this looks like a "workaround" and does not support unbounded, open or closed boundaries as nicely.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/range.html
[2] https://www.sothawo.com/2020/09/how-to-use-elasticsearchs-range-types-with-spring-data-elasticsearch/

### PR DESCRIPTION
**Pull request by xhaggi**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.
  com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Closes #1862

### CODE REVIEW

File: src/test/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverterUnitTests.java [hunk 1]
```diff
@@ -882,6 +889,175 @@ void shouldWriteNullValueIfConfigured() throws JSONException {
...
+			// given
+			Document source = Document.parse(JSON);
+
+			// when
+			RangeEntity entity = mappingElasticsearchConverter.read(RangeEntity.class, source);
+
+			// then
+			assertThat(entity) //
+					.isNotNull() //
+					.satisfies(e -> {
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverterUnitTests.java:934]: I learned something new! Did not know about `satisifes`.

File: src/test/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverterUnitTests.java [hunk 2]
```diff
@@ -882,6 +889,175 @@ void shouldWriteNullValueIfConfigured() throws JSONException {
 		assertEquals(expected, document.toJson(), false);
 	}
 
+	@Nested
+	class RangeTests {
+
+		static final String JSON = "{\"_class\":\"org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverterUnitTests$RangeTests$RangeEntity\",\"integerRange\":{\"lt\":\"1\",\"gt\":\"10\"},\"floatRange\":{\"lte\":\"1.2\",\"gte\":\"2.5\"},\"longRange\":{\"lt\":\"2\",\"gte\":\"5\"},\"doubleRange\":{\"lte\":\"3.2\",\"gt\":\"7.4\"},\"dateRange\":{\"lte\":\"1970-01-01T00:00:00.000Z\",\"gte\":\"1970-01-01T01:00:00.000Z\"},\"localDateRange\":{\"lte\":\"2021-07-06\"},\"localTimeRange\":{\"lte\":\"00:30:00.000\",\"gt\":\"02:30:00.000\"},\"localDateTimeRange\":{\"lt\":\"2021-01-01T00:30:00.000\",\"gt\":\"2021-01-01T02:30:00.000\"},\"offsetTimeRange\":{\"lte\":\"00:30:00.000+02:00\",\"gt\":\"02:30:00.000+02:00\"},\"zonedDateTimeRange\":{\"lte\":\"2021-01-01T00:30:00.000+02:00\",\"gte\":\"2021-01-01T00:30:00.000+02:00\"}}";
```
Comment (sothawo) [inline src/test/java/org/springframework/data/elasticsearch/core/convert/MappingElasticsearchConverterUnitTests.java:895]: I'd prefer to have that in a more readable multiline format

**Review by sothawo — CHANGES_REQUESTED** [review]
Nice work; i left a couple of comments.

What do you think about adding default range types like `class IntegerRange extends Range<Integer>{}`?

About the IPs - would be great to have that as well. I'll see if I find something useful

**Review by xhaggi — COMMENTED** [review]
>Nice work; i left a couple of comments.
Thank you, I have addressed your comments.

>What do you think about adding default range types like class IntegerRange extends Range<Integer>{}?
I have no strong opinion on that. Writing Range<Integer> or IntegerRange makes no difference to me.

>About the IPs - would be great to have that as well.
Sure, but IMO we could add it later.

**Review by sothawo — APPROVED** [review]

**Pull request merged**